xarray/core/indexing.py

```diff
@@ -434,7 +492,84 @@ def __setitem__(self, key, value):
         self.array[key] = value
 
 
-def orthogonally_indexable(array):
+def unbroadcast_indexes(key, shape):
+    """
+    Convert broadcasted indexers to orthogonal indexers.
+    If there is no valid mapping, raises IndexError.
+
+    key is usually generated by Variable._broadcast_indexes.
+
+    key: tuple of np.ndarray, slice, integer
+    shape: shape of array
+    """
+    key = expanded_indexer(key, len(shape))
+
+    if all(isinstance(k, integer_types + (slice,)) for k in key):
+        # basic indexing
+        return key
+
+    # Here, we need to distinguish integer and 1-element array
```

Comment:
Indeed, this logic is quite tricky.

It's worth taking a step back and considering why we need this logic:
- Various backends don't support full vectorized indexing, but are OK with mixed slices/integers with a single 1D arrays. This includes the important cases of dask and h5py.
- We currently normalize indexes with mixed slices/arrays to always use vectorized indexing. We do this so we can make use of xarray's broadcasting.

I wonder if instead of converting `int`/`slice` objects into `xarray.Variable` object for broadcasting, we could keep track of the original objects in some way.

Another choice would be to keep a code path for existing, orthogonal indexing, and still use in cases where it's equivalent to vectorized indexing.